src/ERC20Reserve.sol

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: GPL-3.0-only
+pragma solidity ^0.8.7;
+
+import {Ownable} from "openzeppelin-contracts/access/Ownable.sol";
+import {IERC20} from "openzeppelin-contracts/token/ERC20/IERC20.sol";
+
+interface IERC20Reserve {
+    function withdraw(uint256 amt) external;
+
+    function deposit(uint256 amt) external;
+}
+
+contract ERC20Reserve is IERC20Reserve, Ownable {
+    IERC20 public immutable erc20;
+    address public user;
+    uint256 public balance;
+
+    event Withdrawn(address to, uint256 amt);
+    event Deposited(address from, uint256 amt);
+    event ForceWithdrawn(address to, uint256 amt);
+    event UserSet(address oldUser, address newUser);
+
+    constructor(
+        IERC20 _erc20,
+        address owner,
+        address _user
+    ) {
+        erc20 = _erc20;
+        transferOwnership(owner);
+        setUser(_user);
+    }
+
+    modifier onlyUser() {
+        require(_msgSender() == user, "Reserve: caller is not the user");
+        _;
+    }
+
+    function withdraw(uint256 amt) public override onlyUser {
+        require(balance >= amt, "Reserve: withdrawal over balance");
+        balance -= amt;
+        emit Withdrawn(_msgSender(), amt);
+        require(erc20.transfer(_msgSender(), amt), "Reserve: transfer failed");
+    }
+
+    function deposit(uint256 amt) public override onlyUser {
+        balance += amt;
+        emit Deposited(_msgSender(), amt);
+        require(erc20.transferFrom(_msgSender(), address(this), amt), "Reserve: transfer failed");
+    }
+
+    function forceWithdraw(uint256 amt) public onlyOwner {
```

Comment:
I am wondering if we should just have one `withdraw` method which can be used by both `user` and `owner`. Since the user is a smart contract which doesn't have the calls for `setUser` implemented, we could also just use a `mapping` with one auth type.

No, strong opinion however. I think owner should be able to take the funds without first assigning the user role to itself.

Reply:
Ahh, now i saw the difference the `forceWithdraw` allows you to take all of it. (if we have more the balance)